### PR TITLE
[SPARK-23555][PYTHON] Add BinaryType support for Arrow in Python

### PR DESCRIPTION
**Pull request by BryanCutler**
## What changes were proposed in this pull request?

Adding `BinaryType` support for Arrow in pyspark, conditional on using pyarrow >= 0.10.0. Earlier versions will continue to raise a TypeError.

## How was this patch tested?

Additional unit tests in pyspark for code paths that use Arrow for createDataFrame, toPandas, and scalar pandas_udfs.

### CODE REVIEW

**Comment by BryanCutler** [comment]
This is a WIP as some issues need to be worked out on the Arrow side and need tests for pandas_udfs.  Currently get the following error when converting from pandas:
```
File "/home/bryan/git/spark/python/pyspark/serializers.py", line 237, in create_array
    return pa.Array.from_pandas(s, mask=mask, type=t)
  File "array.pxi", line 335, in pyarrow.lib.Array.from_pandas
  File "array.pxi", line 170, in pyarrow.lib.array
  File "array.pxi", line 70, in pyarrow.lib._ndarray_to_array
  File "error.pxi", line 85, in pyarrow.lib.check_status
ArrowNotImplementedError: No cast implemented from binary to binary
```
The corresponding JIRA is https://issues.apache.org/jira/browse/ARROW-2141

File: python/pyspark/sql/tests.py [hunk 1]
```diff
@@ -4331,13 +4354,22 @@ def test_createDataFrame_fallback_enabled(self):
...
             with self.assertRaisesRegexp(TypeError, 'Unsupported type'):
                 self.spark.createDataFrame(
                     pd.DataFrame([[{u'a': 1}]]), "a: map<string, int>")
 
+        # TODO: remove BinaryType check once minimum pyarrow version is 0.10.0
+        if LooseVersion(pa.__version__) < LooseVersion("0.10.0"):
+            with QuietTest(self.sc):
+                with self.assertRaisesRegexp(TypeError, 'Unsupported type.*BinaryType'):
+                    self.spark.createDataFrame(
+                        pd.DataFrame([[{'a': b'aaa'}]]), "a: binary")
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests.py]: @BryanCutler, BTW, what kind of data can be binary? `bytearray` and `byte` (`str` in Python 2)?

**Review by HyukjinKwon — APPROVED** [review]
LGTM. one trivial question.

File: python/pyspark/sql/tests.py (hunk 1 above)
Comment (BryanCutler) [reply python/pyspark/sql/tests.py]: In general, Spark only accepts `bytearray` and pyarrow can accept `bytearray` or `bytes`. In pyarrow < 0.10.0 bytearrays weren't supported though, which is why I used `bytes` for the pandas pdf here.

**Comment by shaneknapp** [comment]
https://rise.cs.berkeley.edu/jenkins/job/ubuntuSparkPRB/79/

**Comment by shaneknapp** [comment]
https://rise.cs.berkeley.edu/jenkins/job/ubuntuSparkPRB/80/

**Comment by shaneknapp** [comment]
i've been having trouble getting this build to pass, and i'm testing a few other modules on top of yours.  if i have problems, i'll revert the extraneous changes and build yours.

**Comment by shaneknapp** [comment]
still having problems getting this to pass:
```
[error] (sql-kafka-0-10/test:test) sbt.TestsFailedException: Tests unsuccessful
[error] (core/test:test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 14786 s, completed Aug 16, 2018 4:05:09 PM
```

i rebased upstream changes from the main spark repo on my fork and launched another build.  in ~5 hours we'll know how it went.  :\

if this fails tonite, i'll figure out a hacky way to test this tomorrow morning.

**Comment by BryanCutler** [comment]
Awesome, thanks for running that @shaneknapp !!  I'll merge this later today if no further comments.

File: python/pyspark/sql/types.py [hunk 2]
```diff
@@ -1597,6 +1598,12 @@ def to_arrow_type(dt):
         arrow_type = pa.decimal128(dt.precision, dt.scale)
     elif type(dt) == StringType:
         arrow_type = pa.string()
+    elif type(dt) == BinaryType:
+        # TODO: remove version check once minimum pyarrow version is 0.10.0
+        if LooseVersion(pa.__version__) < LooseVersion("0.10.0"):
+            raise TypeError("Unsupported type in conversion to Arrow: " + str(dt) +
```
Comment (gatorsmile) [inline python/pyspark/sql/types.py]: Could you document all the features that require pyarrow 0.10? 

Thanks!

**Pull request closed without merging**